examples/ul2/src/utils/hf_prefixlm_converter.py

```diff
@@ -0,0 +1,769 @@
+                                      input_ids: torch.LongTensor,
+                                      past: Optional[torch.Tensor] = None,
+                                      attention_mask: Optional[
+                                          torch.Tensor] = None,
+                                      **kwargs) -> dict:
+        # assert kwargs.get("use_cache"), "Make sure to set `use_cache` to `True`."
+        # only last token for input_ids if past is not None
+        if past:
+            input_ids = input_ids[:, -1].unsqueeze(-1)
+            bidirectional_mask = None  # We can turn off bidirectional masking after the prefix has been encoded into `past`
+
+            # the cache may be in the stardard format (e.g. in contrastive search), convert to bloom's format if needed
+            if past[0][0].shape[0] == input_ids.shape[0]:
+                past = self._convert_to_bloom_cache(past)
+
+        else:
+            # If we're here, `input_ids` contains the prefix. Encode it with bidirectional attention.
+            bidirectional_mask = torch.ones_like(input_ids)
+
+        return {
+            'input_ids': input_ids,
+            'past_key_values': past,
+            # "use_cache": kwargs.get("use_cache"),
+            'use_cache':
+                True,  # Requires this. TODO(Alex): Confirm this supports other decoding strategies.
+            'attention_mask': attention_mask,
+            'bidirectional_mask': bidirectional_mask,
+        }
+
+    # Register the new `forward` and `prepare_inputs_for_generation` methods with the model
+    setattr(model, 'forward', MethodType(forward, model))
+    setattr(model, 'prepare_inputs_for_generation',
+            MethodType(prepare_inputs_for_generation, model))
+
+    # Finally, tag the model so that this conversion cannot happen again.
+    setattr(model, '_prefix_lm_converted', True)
+    return model
+
+
+def _convert_opt_causal_lm_to_prefix_lm(
+        model: OPTForCausalLM) -> OPTForCausalLM:
+    """Converts an OPT Causal LM to a Prefix LM.
+
+    Supported HuggingFace model classes:
+        - `OPTForCausalLM`
+
+    See `convert_hf_causal_lm_to_prefix_lm` for more details.
+    """
+    if hasattr(model, '_prefix_lm_converted'):
+        return model
+
+    assert isinstance(model, OPTForCausalLM)
+    assert model.config.add_cross_attention == False, 'Only supports OPT decoder-only models'
+
+    # Rename methods to allow:
+    #  - new `forward` to wrap original `forward`
+    #  - new `generate` to wrap original `generate`
+    setattr(model, '_original_forward', getattr(model, 'forward'))
+    setattr(model, '_original_generate', getattr(model, 'generate'))
+
+    # Modified from transformers.models.bloom.modeling_opt.OPTDecoder._prepare_decoder_attn_mask
+    model.model.decoder.bidirectional_mask = None
+
+    def _prepare_decoder_attention_mask(self, attention_mask, input_shape,
+                                        inputs_embeds, past_key_values_length):
```

Comment:
method -> function?

Reply:
This overwrites a method of the same name in the existing model code, so we're kinda stuck with it being a method, I think.